File: day02/day02.zig
const std = @import("std");

const Range = struct {
    start: i64,
    stop: i64,
};

fn parseInput(allocator: std.mem.Allocator, input: []const u8) !std.ArrayList(Range) {
    var ranges: std.ArrayList(Range) = .empty;

    const trimmed = std.mem.trim(u8, input, " \n\r");
    const without_trailing_comma = if (trimmed.len > 0 and trimmed[trimmed.len - 1] == ',')
        trimmed[0 .. trimmed.len - 1]
    else
        trimmed;

    var it = std.mem.splitScalar(u8, without_trailing_comma, ',');
    while (it.next()) |range_str| {
        var range_it = std.mem.splitScalar(u8, range_str, '-');
        const start_str = range_it.next() orelse continue;
        const stop_str = range_it.next() orelse continue;

        const start = std.fmt.parseInt(i64, start_str, 10) catch continue;
        const stop = std.fmt.parseInt(i64, stop_str, 10) catch continue;

        try ranges.append(allocator, .{ .start = start, .stop = stop });
    }

    return ranges;
}

fn countDigits(n: i64) usize {
    if (n == 0) return 1;
    var num = if (n < 0) -n else n;
    var count: usize = 0;
    while (num > 0) {
        num = @divTrunc(num, 10);
        count += 1;
    }
    return count;
}

fn pow10(exp: usize) i64 {
    var result: i64 = 1;
    for (0..exp) |_| {
        result *= 10;
    }
    return result;
}

fn findInvalidIds(allocator: std.mem.Allocator, start: i64, stop: i64, exactly_two_reps: bool) !std.ArrayList(i64) {
    var result: std.ArrayList(i64) = .empty;

    const start_digits = countDigits(start);
    const stop_digits = countDigits(stop);

    for (start_digits..stop_digits + 1) |digit_count| {
        for (1..digit_count / 2 + 1) |pattern_len| {
            if (digit_count % pattern_len != 0) continue;

            const reps = digit_count / pattern_len;
            if (exactly_two_reps and reps != 2) continue;
            if (!exactly_two_reps and reps < 2) continue;

            const min_pattern: i64 = if (pattern_len == 1) 1 else pow10(pattern_len - 1);
            const max_pattern: i64 = pow10(pattern_len) - 1;

            var pattern = min_pattern;
            while (pattern <= max_pattern) : (pattern += 1) {
                // Build the full number by repeating the pattern
                var full_num: i64 = 0;
                for (0..reps) |_| {
                    full_num = full_num * pow10(pattern_len) + pattern;
                }

                if (full_num >= start and full_num <= stop) {
                    try result.append(allocator, full_num);
                }
            }
        }
    }

    return result;
}

fn solve(allocator: std.mem.Allocator, ranges: []const Range, exactly_two_reps: bool) !i64 {
    var seen = std.AutoHashMap(i64, void).init(allocator);
    defer seen.deinit();

    for (ranges) |range| {
        var invalid_ids = try findInvalidIds(allocator, range.start, range.stop, exactly_two_reps);
        defer invalid_ids.deinit(allocator);

        for (invalid_ids.items) |id| {
            try seen.put(id, {});
        }
    }

    var sum: i64 = 0;
    var iter = seen.keyIterator();
    while (iter.next()) |key| {
        sum += key.*;
    }

    return sum;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const file = try std.fs.cwd().openFile("input.txt", .{});
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    var ranges = try parseInput(allocator, content);
    defer ranges.deinit(allocator);

    const part1 = try solve(allocator, ranges.items, true);
    const part2 = try solve(allocator, ranges.items, false);

    const stdout = std.fs.File.stdout();
    var buf: [64]u8 = undefined;

    var out = std.fmt.bufPrint(&buf, "Part 1: {d}\n", .{part1}) catch unreachable;
    try stdout.writeAll(out);

    out = std.fmt.bufPrint(&buf, "Part 2: {d}\n", .{part2}) catch unreachable;
    try stdout.writeAll(out);
}

File: day02/day02.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Range struct {
	Start int64
	Stop  int64
}

func parseInput(input string) []Range {
	var ranges []Range

	trimmed := strings.TrimSpace(input)
	trimmed = strings.TrimSuffix(trimmed, ",")

	parts := strings.Split(trimmed, ",")
	for _, part := range parts {
		rangeStr := strings.TrimSpace(part)
		if rangeStr == "" {
			continue
		}

		rangeParts := strings.Split(rangeStr, "-")
		if len(rangeParts) != 2 {
			continue
		}

		start, err1 := strconv.ParseInt(rangeParts[0], 10, 64)
		stop, err2 := strconv.ParseInt(rangeParts[1], 10, 64)
		if err1 != nil || err2 != nil {
			continue
		}

		ranges = append(ranges, Range{Start: start, Stop: stop})
	}

	return ranges
}

func countDigits(n int64) int {
	if n == 0 {
		return 1
	}
	count := 0
	for n > 0 {
		n /= 10
		count++
	}
	return count
}

func pow10(exp int) int64 {
	result := int64(1)
	for i := 0; i < exp; i++ {
		result *= 10
	}
	return result
}

func findInvalidIDs(start, stop int64, exactlyTwoReps bool) []int64 {
	var result []int64

	startDigits := countDigits(start)
	stopDigits := countDigits(stop)

	for digitCount := startDigits; digitCount <= stopDigits; digitCount++ {
		for patternLen := 1; patternLen <= digitCount/2; patternLen++ {
			if digitCount%patternLen != 0 {
				continue
			}

			reps := digitCount / patternLen
			if exactlyTwoReps && reps != 2 {
				continue
			}
			if !exactlyTwoReps && reps < 2 {
				continue
			}

			minPattern := int64(1)
			if patternLen > 1 {
				minPattern = pow10(patternLen - 1)
			}
			maxPattern := pow10(patternLen) - 1

			for pattern := minPattern; pattern <= maxPattern; pattern++ {
				// Build the full number by repeating the pattern
				fullNum := int64(0)
				multiplier := pow10(patternLen)
				for i := 0; i < reps; i++ {
					fullNum = fullNum*multiplier + pattern
				}

				if fullNum >= start && fullNum <= stop {
					result = append(result, fullNum)
				}
			}
		}
	}

	return result
}

func solve(ranges []Range, exactlyTwoReps bool) int64 {
	seen := make(map[int64]struct{})

	for _, r := range ranges {
		invalidIDs := findInvalidIDs(r.Start, r.Stop, exactlyTwoReps)
		for _, id := range invalidIDs {
			seen[id] = struct{}{}
		}
	}

	var sum int64
	for id := range seen {
		sum += id
	}

	return sum
}

func main() {
	content, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to read input.txt: %v\n", err)
		os.Exit(1)
	}

	ranges := parseInput(string(content))

	part1 := solve(ranges, true)
	part2 := solve(ranges, false)

	fmt.Printf("Part 1: %d\n", part1)
	fmt.Printf("Part 2: %d\n", part2)
}
